sbi/inference/snpe/snpe_base.py

```diff
@@ -164,6 +176,9 @@ def __call__(
             validation_fraction: The fraction of data to use for validation.
             stop_after_epochs: The number of epochs to wait for improvement on the
                 validation set before terminating training.
+            max_num_epochs: maximal number of epochs to run. If max_num_epochs
+                is reached, we stop training even if the validation loss is still
+                decreasing.
```

Comment:
We should document this well, maybe not here, but in a user guide. But the user should know about this one and should check whether validation loss keeps decreasing. Or we should even add a warning if validation loss is decreasing and we are stopping training because of max_num_epochs?

Reply:
Here's where I can't follow (agree): we are adding a limitation of the training that is a bit arbitrary and might be very painful (a lot of simulations spent on a non-converged training) just to be able to normalize a progress bar. 

Would we have added this in absence of the wish of a progresss bar?

Conversely, what's wrong with reporting progress, without normalization?

I feel I am missing something here... :thinking:

Reply:
I'm still a bit torn as to whether we should use this.
Arguments for using `max_num_epochs`:
- indication with progressbar
- if we have very much training data, we will barely ever overfit. Combine this behavior with a high value of `stop_after_epochs` and you get an infinite training loop

Arguments against using `max_num_epochs`:
- we might underfit. As Jan suggested, we are currently raising a warning if `max_num_epochs` is reached. So at least the user knows about this.

I don't have a strong opinion, but I am leaning towards using it. If you are both against I will just take it out though.

Reply:
I'd like to hear some more opinions on this - in particular, how painful is the underfit? AFAIK we don't have a checkpointing mechanism in place whereby the user could examine the posterior and pass it back saying 'can you work a bit more on this'?

Reply:
In practice, the strategy of monitoring the validation set and stopping if there was no improvement for X subsequent epochs seems to work quite well. Being able to set `max_num_epochs` is a good idea but I'd rather keep this turned off by default

Reply:
So in conclusion, we set it to e.g., `Optional[int] = None` and use unnormalized pbars by default and if it is set to an integer we raise a warning if it is reached and the validation loss is still decreasing i.e., we are probably underfitting?

Reply:
Yes, sounds good